sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileBasedSink.java

```diff
@@ -319,10 +319,6 @@ public void populateDisplayData(DisplayData.Builder builder) {}
               DynamicDestinations.class,
               new TypeVariableExtractor<
                   DynamicDestinations<UserT, DestinationT, OutputT>, DestinationT>() {});
-      checkArgument(
-          descriptor != null,
-          "Unable to infer a coder for DestinationT, "
-              + "please specify it explicitly by overriding getDestinationCoder()");
       return registry.getCoder(descriptor);
```

Comment:
Catch the CannotProvideCoderException and rethrow an exception like you did in AvroIO.